clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -678,16 +678,31 @@ void CodeGenFunction::EmitKernelMetadata(const FunctionDecl *FD,
     std::optional<llvm::APSInt> ZDimVal = A->getZDimVal();
     llvm::SmallVector<llvm::Metadata *, 3> AttrMDArgs;
 
+    llvm::APInt NumDims(32, 1); // X
     // On SYCL target the dimensions are reversed if present.
-    if (ZDimVal)
+    if (ZDimVal) {
       AttrMDArgs.push_back(
           llvm::ConstantAsMetadata::get(Builder.getInt(*ZDimVal)));
-    if (YDimVal)
+      ++NumDims;
+    }
+    if (YDimVal) {
       AttrMDArgs.push_back(
           llvm::ConstantAsMetadata::get(Builder.getInt(*YDimVal)));
+      ++NumDims;
+    }
     AttrMDArgs.push_back(
         llvm::ConstantAsMetadata::get(Builder.getInt(*XDimVal)));
 
+    if (getLangOpts().SYCLIsDevice && (getTarget().getTriple().isAMDGCN() ||
+                                       getTarget().getTriple().isNVPTX())) {
+      for (auto i = NumDims.getZExtValue(); i < 3; ++i)
+        AttrMDArgs.push_back(
```

Comment:
This is for padding with 1?

The code looks ok but I honestly do not remember this attribute's functionality well enough to comment on it. I forget it within a month of understanding it each time someone works on this...sigh

@steffenlarsen IIRC you worked on this most recently. Would you mind taking a look?

Reply:
Ha ha, that makes two of us :)

That's correct, it keeps track of how many dimensions were specified and pads the missing ones with `1`.